developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+
+The SOF VirtIO driver on the guest registers support for that PCI device and
+its .probe() method is called.
+
+The SOF driver on the guest performs partial initialisation, it skips any
+steps, involving communicating with the actual DSP, e.g. sending the firmware,
+booting it, tracing initialisation.
+
+The VirtIO driver on the guest uses a new SOF_IPC_TPLG_VFE_GET VirtIO-specific
+IPC message to request the topology from the host. The host reads the respective
+file from its filesystem, and sends it in multiple chunks to the guest in reply
+to that request.
+
+TODO: define a reliable secure topology selection protocol. Currently the guest
+simply sends a configured ID to the host, upon which the host opens the topology
+file with the same name. Currently this isn't a problem, since so far only a
+single guest is supported. So, the host can verify, that the ID, that the guest
+has sent does indeed match with the guest topology file name. However, when
+support for multiple guests is added, it must be ascertained, that guests only
+receive topology files, that they are allowed to get. At the moment there
+doesn't seem to be a way on the host to uniquely and securely identify guests.
+If there is indeed no such way, our only possibility seems to be using
+cryptography for that.
+
+Guest topology files contain virtual DAI components of type "dai_out" and
+"dai_in." Their data tuples contain a new token of type SOF_TKN_COMP_REF, which
+contains an ID of the mixer or the demux, to which this pipeline should be
+connected.
+
+After obtaining the topology, the guest SOF instance uses a second new
+SOF_IPC_TPLG_VFE_COMP_ID VirtIO-specific IPC message to request its allocated
+component ID base. The host maintains a bitmap of guest component ID ranges to
+guarantee, that component IDs stay unique regardless of the order in which
+guests are brought up and down.
+
+Next the guest proceeds by sending the parsed topology over IPC to the host,
+which is then forwarded to the DSP firmware. The guest uses the .send_msg()
+callback to forward any IPC, including topology related, with no processing at
+all, to the host. The only exception is blocking IPCs when the guest is
+resuming, while the host doesn't need to be resumed (see 3.3 section on "reset
+count" for details).
+
+The host receives the guest topology over the IPC virtual queue. In most cases
+the host IPC handler just forwards IPCs to the DSP, but in some cases it has to
+handle or modify IPC contents locally. One such example is linking the guest and
+host topology fragments together. As mentioned above, guest topology files
+contain a new token, referencing mixer or demux components in the host topology.
+This value is then assigned to the .ref_comp_id field of struct
+sof_ipc_comp_config during topology file parsing and sent by the guest to the
+host. The host then uses that value to modify connection IPC messages from the
+guest before sending them to the DSP.
```

Comment:
do you mean to say that there is a guest-ID to host-ID translation?

Reply:
kind of, yes. Both those component IDs belong to the same range, but the "guest ID" is an ID of a virtual DAI, that is only present on the guest, and that is then replaced with a component ID from the host part of the pipeline, to which the guest part is attached